Source/ParticleInterpolators/ParticleInterpolators.impl.hpp

```diff
@@ -0,0 +1,655 @@
+            start_comp, ncomp, nghost, geom.periodicity());
+
+        const auto plo = geom.ProbLoArray();
+        const auto dxi = geom.InvCellSizeArray();
+
+        for (ParIterType it(*this, lev); it.isValid(); ++it)
+        {
+            auto arrs    = mf[it].const_array();
+            auto &ptile  = this->ParticlesAt(lev, it);
+            auto ptd     = ptile.getParticleTileData();
+            const int np = it.numRealParticles();
+
+            amrex::ParallelFor(
+                np,
+                [=] AMREX_GPU_DEVICE(int ip)
+                {
+                    auto &sp = ptd[ip];
+
+                    amrex::IntVect is_nodal = amrex::IntVect::TheZeroVector();
+                    LagrangeInterpolator<5> interp;
+                    interp.compute_weights(sp, plo, dxi, is_nodal);
+
+                    amrex::ParticleReal vals[ncomp];
+                    interp.interpolate(&arrs, vals, start_comp, ncomp);
+                    for (int k = 0; k < ncomp; ++k)
+                    {
+                        ptd.rdata(k)[ip] = vals[k];
+                    }
+                });
+
+            amrex::Gpu::streamSynchronize();
+        }
+    }
+
+    m_particles_seeded  = true;
+    m_need_redistribute = false;
+}
+
+// mirror of AMRInterpolator::interp(); assembles all particle data and writes
+// parity * value into the query out arrays
+// Throws an error now if use requests interpolation of a derived variable and
+// applies reflective BCs (needs fixing later!!) This is currently done on CPUs;
+// I am not sure if there is an advantage of having any of this on GPUs
+template <int num_components>
+void ParticleInterpolators<num_components>::interp(
+    InterpolationQueryParticle &query, VariableType variable_type)
+{
+    AMREX_ASSERT(m_initialized);
+
+    // if (variable_type == VariableType::derived ||
+    //     variable_type == VariableType::state)
+    // {
+    //     for (int dir = 0; dir < AMREX_SPACEDIM; ++dir)
+    //     {
+    //         if (m_lo_boundary_reflective[dir] ||
+    //         m_hi_boundary_reflective[dir])
+    //         {
+    //             amrex::Abort("Help!!! reflective BCs are implemented, but not
+    //             "
+    //                          "yet tested. Likely to give absolute
+    //                          gibberish!");
+    //         }
+    //     }
+    // }
+
+    // get total query points here
+    int npts = 0;
+    if (amrex::ParallelDescriptor::IOProcessor())
+    {
+        npts = static_cast<int>(query.numPoints()); // only rank 0 touches query
+    }
+    // broadcast
+    amrex::ParallelDescriptor::Bcast(
+        &npts, 1, amrex::ParallelDescriptor::IOProcessorNumber());
+
+    // value_at_point[k][ip], where k in [0..num_components-1]
+    std::vector<std::vector<amrex::Real>> value_at_point(
+        num_components, std::vector<amrex::Real>(npts, 0.0));
+    // a vector to mark which points have values
+    std::vector<int> have(npts, 0);
```

Comment:
Is this necessary other than for debugging? Maybe just conditionally compile all the code that uses it on `AMREX_DEBUG` e.g.
```suggestion
#ifdef AMREX_DEBUG
    std::vector<int> have(npts, 0);
#endif
```